### DISCUSSION TITLE
datarate

### DISCUSSION THREAD

**Discussion by yjr1220** [svpcom/wfb-ng#150] [Questions]
Hello ! my dear!

        I want to know whether the transmission rate can be changed in monitor mode. For example, I can set the rate under 802.11 AC of WiFi card! High speed transmission!  

         best wishes !

**Comment by svpcom** [comment]
Use `mcs_index = value` in config file and use value from http://mcsindex.com/

**Reply by yjr1220** [reply]
thank you for reply!  in you source code,
int main(int argc, char * const *argv)
{
    int opt;
    uint8_t k=8, n=12, radio_port=1;
    int udp_port=5600;

    int bandwidth = 20;
    int short_gi = 0;
    int stbc = 0;
    int ldpc = 0;
    int mcs_index = 1;
 .......
   the bandwidth  can set value 80  or 160 ?  the  mcs_index  can set   value make the datarate change ?  is right ? 
  and the data rate is only related to mcs_index ?
  best wishes !

**Comment by svpcom** [comment]
You can try to use 80 and/or 160 MHz bandwidth, but I'm not sure that current drivers supports injection of VHT packets. Data rate depends from bandwith and mcs index (see table from link above). But I'll recommend to use only 20MHz bandwidth for long-distance transmissions

**Reply by yjr1220** [reply]
ok , thank you very much !  !  !

**Reply by yjr1220** [reply]
hi svpcom !
          injection of VHT packets need the firmware  of hardware support ?  

 best wishes !

**Reply by svpcom** [reply]
It already supported by driver, but you you need to use different radiotap header contents on wfb_tx and rx side. See for details:
https://github.com/svpcom/rtl8812au/blob/e9fbf5c051453941bbc029810b893a6c010714e6/core/rtw_xmit.c#L4646

**Reply by yjr1220** [reply]
thank you !  !  !

**Reply by yjr1220** [reply]
Hi svpcom

      I want to know that data inject under monitor mode must be support by firmware of the wifi chip ?  if not the data packet can not transmit !  is right ? 

       I have a wifi card  rtl8822cs . It can set mode monitor ,but can not transmit data.  so I think  it has not be support by the firmware of the wifi chip.
best wishes!!!

**Reply by svpcom** [reply]
Packet injection should be supported **both** by firmware and driver. For example rtl8812au supports injection by firmware, but stock driver doesn't support it (there are special driver with injecton https://github.com/svpcom/rtl8812au).
Most of wifi cards doesn't support injection both by firmware and driver, but supports monitor mode.

**Reply by yjr1220** [reply]
Oh! I understand.  thank you for my reply !

**Comment by yjr1220** [comment]
hi svpcom !
                  How can I fix the frequency in some band ?  example ,  I want to use  40 channel  80M bandwidth . 
I use VHT injection package, must set the bandwidth right? I am not sure how to fix the mode at 80211.ac under the monitor mode.

best wishes !!!

**Reply by svpcom** [reply]
Channel selected by external program (iw) at the service start. Bandwidth set in every injected packet (different packets can have different bandwidth and/or MCS index)

**Reply by yjr1220** [reply]
Hi svpcom !

That is to say, there is no need to set bandwidth or hardware mode( B, G, N, AC... And so on). I just set the channel with the (IW) tool, and then set the bandwidth and MCS in your program, right? 
your program is：
{
int bandwidth = 20(40,80,160 and so on);
int short_gi = 0;
int stbc = 0;
int ldpc = 0;
int mcs_index = 1(2,3,4,5,6...);
........
}
When I send the inject package it will be sent at the bandwidth and MCS rate set in the program, right?

best wishes !!!

**Reply by svpcom** [reply]
yes

**Reply by yjr1220** [reply]
ok , thank you very much!

**Reply by yjr1220** [reply]
Hi svpcom !

      With your help, I changed the ratio head part into the VHT rate by modifying your code. Debugging the driver shows that the following code does execute. 
#ifdef CONFIG_80211AC_VHT
			case IEEE80211_RADIOTAP_VHT: 
			{
				unsigned int mcs, nss;
				u8 known = iterator.this_arg[1];
				u8 flags = iterator.this_arg[2];
                                nss = iterator.this_arg[4] & 0x0F; // Number of spatial streams
                               ......
                                if (nss > 0) 
                                {
					if (nss > 4)    nss = 4;
					mcs = (iterator.this_arg[4]>>4) & 0x0F; // MCS rate index
					if (mcs > 8)   mcs = 9;
					pattrib->rate = MGN_VHT1SS_MCS0 + ((nss-1)*10 + mcs);
					if (iterator.this_arg[8] & IEEE80211_RADIOTAP_CODING_LDPC_USER0)
						pattrib->ldpc = _TRUE;
				}              
			} 
			break;
#endif
However, the rate did not change, I tested only 40Mbps. I don't know where the problem is. Could you please help analyze it? The modifications are as follows.
static uint8_t radiotap_header[]  __attribute__((unused)) = {
    0x00, 0x00,             // <-- radiotap version
    0x16, 0x00,             // <- radiotap header length
    0x00, 0x80, 0x20, 0x00, // <-- radiotap present flags:  RADIOTAP_TX_FLAGS + RADIOTAP_VHT
    0x08, 0x00,             // RADIOTAP_F_TX_NOACK
    0x00, 0x45, 0x05, 0x04, 0x82,0x00,0x00,0x00,0x01,0x00,0x00,0x00 //VHT u16 known, u8 flags, u8 bandwidth, u8 cs_nss[4], u8 coding, u8 group_id, u16 partial_aid
  //  MCS_KNOWN ,             // bitmap
  //  0x00,                   // flags
  //  0x00                    // mcs_index
};
     When I measured  with the spectrograph, the wifi bandwidth was still 20Mz and it didn't change. 
     thank you !!!

**Reply by yjr1220** [reply]
Hi svpcom !    So far, I've been working on this problem, but the rate has remained around 30Mbps. Does firmware already limit the transmit rate in monitor mode? best wishes !!!

**Comment by svpcom** [comment]
Try to capture them via tcpdump or wireshark - are they really transmitted as VHT frames?

**Reply by yjr1220** [reply]
ok, i'll try!

**Reply by yjr1220** [reply]
Hi svpcom , I don't know how to test !  tcpdump -i wlan0  right ?
